modules: tf-m: rework output file generation

JordanYates opened the pull request:
Stop `mergehex.py` argument `--output-bin` from consuming the first trailing argument that should have been in `input_files`.

`S_NS_CONFIRMED_HEX_FILE` was never generating a confirmed file, just the same file contents as `S_NS_HEX_FILE`. Since no logic needs a confirmed merge of `tfm_s.hex` and `zephyr.hex`, just remove the logic instead of fixing it.

Generate a binary version of `tfm_s_zephyr_ns_signed.hex` with objcopy. This file is valid for performing OTA upgrades, unlike `tfm_merged.bin`, which now contains BL2.

Generate `tfm_merged.bin` from `tfm_merged.hex` with objcopy.

Fixes #98728

After this PR, the correct file to use for OTA upgrades with `CONFIG_TFM_MCUBOOT_IMAGE_NUMBER=1` is `tfm_s_zephyr_ns_signed.bin`, not `tfm_merged.bin`.

JordanYates committed:
scripts: build: mergehex: fix `--output-bin`

Stop `--output-bin` from consuming the first trailing argument that
should have been in `input_files`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
modules: tf-m: remove `S_NS_CONFIRMED_HEX_FILE`

`S_NS_CONFIRMED_HEX_FILE` was never generating a confirmed file, just
the same file contents as `S_NS_HEX_FILE`. Since no logic needs a
confirmed merge of `tfm_s.hex` and `zephyr.hex`, just remove the logic
instead of fixing it.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
modules: tf-m: generate `tfm_s_zephyr_ns_signed.bin`

Generate a binary version of `tfm_s_zephyr_ns_signed.hex` with objcopy.
This file is valid for performing OTA upgrades, unlike `tfm_merged.bin`,
which contains BL2.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
modules: tf-m: generate `tfm_merged.bin` from `tfm_merged.hex`

Since `tfm_merged.bin` now contains BL2, it can only be used for the
same purposes as `tfm_merged.hex` (intial firmware loading). Therefore
it should be using the confirmed images that `tfm_merged.hex` does.

Since the only difference between the two files with that change is now
the output format, we can directly generate `tfm_merged.bin` from
`tfm_merged.hex` with `objcopy` instead of going through `mergehex.py`.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
modules: tf-m: fix CMake formatting

Fix formatting errors that cause CI to complain.

Signed-off-by: Jordan Yates <jordan@embeint.com>

nandojve commented:
Hi @JordanYates ,

I did not look in deep your changes but I would like to reinfoce:

1- The image that is flashed in production and in development (HEX FILE) MUST be confirmed.
  - This is necessary to make sure that TF-M state machine will work correctly.
  - Even if people say _it is necessary to have a confirmed image_ to run I can argue that in the same way having confirmed do not affect. The main difference is that a confirmed image means (in terms of FWU) that it is a sane image.
2- The image that is used in FOTA (BIN FILE) MUST NOT be confirmed. User APP must confirm the image and it should be part of the TF-M FWU procedure.

That said, my original intention was (following your recommendation in the original PR) create valid names to differentiate the confirmed vs non-confirmed image. First should be used in production and second in FOTA.

In my eyes, at least in your last commit, it seems that you are providing the binary confirmed - which violates [2].

So, if there are different use cases that requires the BIN to be already confirmed we need to provide a way to fulfill that requirement in all scenarios.

JordanYates commented:
> I did not look in deep your changes but I would like to reinfoce:

This is a frustrating comment. If you haven't looked at the details why bring up points that you have no idea whether they are relevant or not. It just wastes everyones time.

> 1- The image that is flashed in production and in development (HEX FILE) MUST be confirmed.

Great, absolutely nothing has changed with respect to `tfm_merged.hex`, and `tfm_merged.bin` can now actually be used for that purpose.

> 2- The image that is used in FOTA (BIN FILE) MUST NOT be confirmed.

Great, that is the purpose of `tfm_s_zephyr_ns_signed.bin`, which is not confirmed.

> That said, my original intention was (following your recommendation in the original PR) create valid names to differentiate the confirmed vs non-confirmed image

Not to be rude, but it sounds like you understood my recommendation and then ignored it. `file_name_x.hex` and `file_name_x.bin` should not contain fundamentally different file contents, which is what ended up being merged (as `tfm_merged.hex` and `tfm_merged.bin`).

> In my eyes, at least in your last commit, it seems that you are providing the binary confirmed - which violates [2].

`tfm_merged.bin` SHOULD have the exact same file contents as `tfm_merged.hex`. Disagreeing with that statement is the same as saying `zephyr.hex` should contain a MCUboot image while `zephyr.bin` should not. It doesn't make any sense.

The correct file to use for OTA updates (As explicitly said in the PR description) is `tfm_s_zephyr_ns_signed.bin`.

nandojve commented:
@JordanYates ,

> > I did not look in deep your changes but I would like to reinfoce:
> 
> This is a frustrating comment. If you haven't looked at the details why bring up points that you have no idea whether they are relevant or not. It just wastes everyones time.

I wish I had the proper amount of time to proper evaluate and you are talking about _frustration_. _Frustration_ is working for months to get issues fixed and the PR got merged in the last day because someone don't came back. When I had a setup and time to evaluate where were you? Now in the RC-2 you want people be available to you as nothing happened. You could at least tag me to double check, which in my humble opinion, is the minimal to be professional. But thank you for clarify my concerns.

I was looking in the docs and they need to be changed to not create confusion.

<img width="1142" height="379" alt="image" src="https://github.com/user-attachments/assets/00624610-cb6d-4a77-99fb-0a544627b732" />

I think it will be necessary to state what are the new artifacts to make clear to everyone. Proper naming convention could be necessary to easy follow the main idea. You want to have BIN files generated as same as HEX, I think this should be conditioned to the Kconfig option that states to generate bin files copies.

JordanYates committed:
doc: tfm: document output files

Document the useful output files that exist, which files they are
constructed from, and what they can be used for.

Update other sections that are no longer correct with changes.

Signed-off-by: Jordan Yates <jordan@embeint.com>

nordicjm reviewed (commented):
it should be noted that what happens in this file is (and was) wrong, there are Kconfigs for specifying the output file formats, `CONFIG_BUILD_OUTPUT_HEX` and `CONFIG_BUILD_OUTPUT_BIN`, neither of which seem to be checked or used here

tejlmand reviewed (approved):
Approving this PR as it does fixes issues related to earlier merged PR.

That said, I do think its unfortunate that bin output is depending on hex output, but as TF-M enforces hex output, then this is not an issue in practice.

nandojve reviewed (approved):
Thank you for the fixing @JordanYates
